mlflow/types/llm.py

```diff
@@ -0,0 +1,306 @@
+        self._validate_field("temperature", float, True)
+        self._validate_field("max_tokens", int, False)
+        self._validate_list("stop", str, False)
+        self._validate_field("n", int, True)
+        self._validate_field("stream", bool, True)
+
+
+@dataclass()
+class ChatRequest(ChatParams):
+    """
+    Format of the request object expected by the chat endpoint.
+
+    :param messages:    A list of :py:class:`ChatMessage` that will be passed to the model.
+                        **Optional**, defaults to empty list (``[]``)
+    :type messages:     List[:py:class:`ChatMessage`]
+    :param temperature: A param used to control randomness and creativity during inference.
+                        **Optional**, defaults to ``1.0``
+    :type temperature:  float
+    :param max_tokens:  The maximum number of new tokens to generate.
+                        **Optional**, defaults to ``None`` (unlimited)
+    :type max_tokens:   int
+    :param stop:        A list of tokens at which to stop generation.
+                        **Optional**, defaults to ``None``
+    :type stop:         List[str]
+    :param n:           The number of responses to generate. **Optional**, defaults to ``1``
+    :type n:            int
+    :param stream:      Whether to stream back responses as they are generated.
+                        **Optional**, defaults to ``False``
+    :type stream:       bool
+    """
+
+    messages: List[ChatMessage] = field(default_factory=list)
+
+    def __post_init__(self):
+        self._convert_dataclass_list("messages", ChatMessage)
+        super().__post_init__()
+
+
+@dataclass
+class ChatChoice(_BaseDataclass):
+    """
+    A single chat response generated by the model.
+
+    :param index:         The index of the response in the list of responses.
+    :type index:          int
+    :param message:       The message that was generated.
+    :type message:        :py:class:`ChatMessage`
+    :param finish_reason: The reason why generation stopped.
+    :type finish_reason:  str
+    """
+
+    index: int
+    message: ChatMessage
+    finish_reason: str
+
+    def __post_init__(self):
+        self._validate_field("index", int, True)
+        self._validate_field("finish_reason", str, True)
+        if not isinstance(self.message, ChatMessage):
+            self.message = ChatMessage(**self.message)
+
+
+@dataclass
+class TokenUsageStats(_BaseDataclass):
+    """
+    Stats about the number of tokens used during inference.
+
+    :param prompt_tokens:     The number of tokens in the prompt.
+    :type prompt_tokens:      int
+    :param completion_tokens: The number of tokens in the generated completion.
+    :type completion_tokens:  int
+    :param total_tokens:      The total number of tokens used.
+    :type total_tokens:       int
+    """
+
+    prompt_tokens: int
+    completion_tokens: int
+    total_tokens: int
+
+    def __post_init__(self):
+        self._validate_field("prompt_tokens", int, True)
```

Comment:
Is defining this as a required set of fields going to preclude using this interface in transformers?

Reply:
i don't think it will preclude that, because we can generate these stats automatically for the user using the transformer's tokenizer. however, i can make it not required if it's a concern! it was unclear from the spec which fields are required and not.